### PR TITLE
Issue #1048: Take data version from request param if present.

### LINKED ISSUE #1048: Add feature to specify the data version in the request
This can then be used to preview a catalog data version.

Cookie and/or query parameter.

### PR DESCRIPTION
**Pull request by Vinai**
Closes #1048

### CODE REVIEW

File: src/Context/DataVersion/ContextVersion.php [hunk 1]
```diff
@@ -22,13 +26,50 @@ public function __construct(DataVersion $dataVersion)
...
+    public function getValue(array $inputDataSet): string
     {
-        return (string) ($inputDataSet[DataVersion::CONTEXT_CODE] ?? $this->dataVersion);
+        if ($this->hasVersion($inputDataSet)) {
+            return (string) $this->getVersion($inputDataSet);
+        }
+        if ($this->hasVersionRequestParam($inputDataSet)) {
+            return (string) $this->getVersionRequestParamValue($inputDataSet);
+        }
+        return (string) $this->dataVersion;
```
Comment (tim-bezhashvyly) [inline src/Context/DataVersion/ContextVersion.php]: Also not quite clear. Reads like it is a default version.

File: src/Context/DataVersion/ContextVersion.php [hunk 2]
```diff
@@ -22,13 +26,50 @@ public function __construct(DataVersion $dataVersion)
...
+    private function hasVersionRequestParam(array $inputDataSet): bool
+    {
+        return $this->hasRequest($inputDataSet) &&
+               $this->getRequest($inputDataSet)->hasQueryParameter(self::DATA_VERSION_REQUEST_PARAM);
+    }
+
+    private function getVersionRequestParamValue(array $inputDataSet)
+    {
+        return $this->getRequest($inputDataSet)->getQueryParameter(self::DATA_VERSION_REQUEST_PARAM);
+    }
```
Comment (tim-bezhashvyly) [inline src/Context/DataVersion/ContextVersion.php]: Feel free to ignore all comments regarding this class if you think they make no sense. Maybe I'm just too off the track.

File: src/Import/RestApi/CatalogImportApiV1PutRequestHandler.php [hunk 3]
```diff
@@ -5,12 +5,30 @@
...
+        DataPoolReader $dataPoolReader
+    ) {
+        parent::__construct($importDirectoryPath, $commandQueue, $logger);
+        $this->dataPoolReader = $dataPoolReader;
+    }
+
     final protected function createDataVersion(HttpRequest $request): DataVersion
     {
-        return DataVersion::fromVersionString('-1');
+        return DataVersion::fromVersionString($this->dataPoolReader->getCurrentDataVersion());
```
Comment (tim-bezhashvyly) [inline src/Import/RestApi/CatalogImportApiV1PutRequestHandler.php]: Finally. 👍

File: src/Context/DataVersion/ContextVersion.php [hunk 4]
```diff
@@ -22,13 +26,50 @@ public function __construct(DataVersion $dataVersion)
      * @param mixed[] $inputDataSet
      * @return string
      */
-    public function getValue(array $inputDataSet) : string
+    public function getValue(array $inputDataSet): string
     {
-        return (string) ($inputDataSet[DataVersion::CONTEXT_CODE] ?? $this->dataVersion);
+        if ($this->hasVersion($inputDataSet)) {
```
Comment (Vinai) [inline src/Context/DataVersion/ContextVersion.php]: Renamed to `hasVersionInDataSet()`

File: src/Context/DataVersion/ContextVersion.php [hunk 5]
```diff
@@ -22,13 +26,50 @@ public function __construct(DataVersion $dataVersion)
...
+            return (string) $this->getVersion($inputDataSet);
```
Comment (Vinai) [inline src/Context/DataVersion/ContextVersion.php]: Renamed to `getVersionFromDataSet()`

File: src/Context/DataVersion/ContextVersion.php (hunk 1 above)
Comment (Vinai) [inline src/Context/DataVersion/ContextVersion.php]: That's good then because that's what it is.

**Comment by Vinai** [comment]
Ready for next review.

**Pull request merged**